### Подготовка

#### Скачивание библиотек и датасета

Скачаем необходимые библиотеки

In [1]:
!pip install -q optuna optuna-integration
import optuna

from optuna.visualization import (
    plot_contour,
    plot_slice,
    plot_optimization_history,
    plot_parallel_coordinate,
    plot_param_importances
)

import pandas as pd

from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import recall_score
from sklearn.datasets import load_breast_cancer

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 364.4/364.4 kB 7.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 97.4/97.4 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 233.5/233.5 kB 14.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.6/78.6 kB 2.4 MB/s eta 0:00:00


В качетве датасета будем использовать медицинские показания жителей американского штата Висконсин, целевая переменная - есть ли у человека рак молочной железы

Классификатором будет выступать метод k ближайших соседей

In [2]:
dataset = load_breast_cancer()
classifier = KNeighborsClassifier()

#### Логирование в PostgreSQL

**Внимание**: данный ноутбук нужно запускать в Jupyter Notebook, Google Colab **НЕ** подойдёт, так как для база данных поднимается на локалхосте

С помощью Doсker будет подниматься локальная база данных, для этого надо прописать следующую команду:

``docker run --name optuna-docker-17.1 -p 5432:5432 -e POSTGRES_USER=user -e POSTGRES_PASSWORD=password -e POSTGRES_DB=optuna_db -d postgres:17.1``

После выполнения процедур подбора гиперпараметров, если вы хотите получить доступ к базе данных, то нужно прописать следующую команду:

``docker exec -it optuna-docker-17.1 psql -U user -W optuna_db``, а дальше прописать пароль ``password``

Для остановки контейнера нужно прописать следующую команду:

``docker stop optuna-docker-17.1``

А для удалени контейнера такую:

``docker rm optuna-docker-17.1``

Эта ссылка нужна для подключения к базе данных

In [3]:
storage_url = "postgresql://user:password@localhost:5432/optuna_db"

#### Информация об оптимизируемых гиперпарамерах

Оптимизируемые гиперпараметры:
- Количество ближайших соседей (равномерное целочисленное распределение от 1 до 100 по логарифмической шкале)
- Веса соседей (категориальное распределение, состоящее из 2 значений - "невзвешенные соседи" и "взвешенные соседи")
- Параметр p в метрике Минковского (целочисленное равномерное распределение от 1 до 20)

In [4]:
params = {
    'n_neighbors': optuna.distributions.IntDistribution(1, 100, log=True),
    'weights': optuna.distributions.CategoricalDistribution(['uniform', 'distance']),
    'p': optuna.distributions.IntDistribution(1, 20)
}

Напишем функцию, которая принимает следующие аргументы:
- Класификатор
- Тренировочые данные
- Количество эпох
- Оптимизируемые гиперпараметры
- Оптимизируемые через гиперпараметры **метрику**
- Имя проедуры оптимизации
- "Направление" обучение (минимизация или максимизация метрики)
- Сэмплер
- Прунер

В самой функции будет будет происходить подбор гиперпараметров с помощью класса OptunaSearchCV, используя предыдущее параметры. Будет использоваться процедура кросс-валидации, причём оценука в логи будет выводиться для "валидационной части" кросс-валидации

Возвращать функция будет объект класса OptunaSearchCV

In [10]:
def optimize(
    classifier, X, y, params, study_name='New study', scoring='recall', direction='maximize',
    n_trials=50, sampler=None, pruner=None
):
    optuna_search = optuna.integration.OptunaSearchCV(
        classifier, params, n_trials=n_trials, scoring=scoring,
        study=optuna.create_study(
            storage=storage_url, direction=direction, study_name=study_name,
            sampler=sampler, pruner=pruner
        )
    )
    optuna_search.fit(X, y)
    return optuna_search

#### Метрика

В качестве метрики будем использовать recall, так как самое страшное для нас - отнести больного человека к здоровому, значит нужно максимально понизить False Negative, а чем выше значение метрики Recall при зафиксированном True Positive, тем меньше значение False Negative

#### Прунеры

В качестве прунеров будем использовать следующие 2:
- Процентильный (оставляются только какой-то процент лучших попыток)
- Пороговый (отсекаются попытки, у которых значение метрики слишком низкое или высокое)

При этом сделем так, чтобы первые 10 попыток отечения попыток не проводилось, и прии этом отсечение будет проводитья через каждые 5 эпох

In [5]:
percentile_pruner = optuna.pruners.PercentilePruner(25, n_startup_trials=10, interval_steps=5)  # Будем оставлять только четверть лучших попыток
threshold_pruner = optuna.pruners.ThresholdPruner(
    lower=0.9, upper=0.985, n_warmup_steps=10, interval_steps=5
)  # Если значение метрики recall меньше 0.9 или больше 0.985, то такая попытка будет отсекаться

#### Сэмплеры

Будем исполодьзовать слеюущие сэмплеры:
- Случайный
- TPE (Tree-structured Parzen Estimator)
- GPS (Gaussian process-based Bayesian optimization)

In [6]:
random_sampler = optuna.samplers.RandomSampler()
TPE_sampler = optuna.samplers.TPESampler()
GPS_sampler = optuna.samplers.GPSampler()

<ipython-input-6-08946e42a961>:3: ExperimentalWarning: GPSampler is experimental (supported from v3.6.0). The interface can change in the future.
  GPS_sampler = optuna.samplers.GPSampler()


### Обучение

In [11]:
%%time
search = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier'
)

search_percentile_random = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier, percentile pruner, random sampler',
    pruner=percentile_pruner, sampler=random_sampler
)

search_percentile_TPE = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier, percentile pruner, TPE sampler',
    pruner=percentile_pruner, sampler=TPE_sampler
)

search_percentile_GPS = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier, percentile pruner, GPS sampler',
    pruner=percentile_pruner, sampler=GPS_sampler
)

search_threshold_random = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier, threshold pruner, random sampler',
    pruner=threshold_pruner, sampler=random_sampler
)

search_threshold_TPE = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier, threshold pruner, TPE sampler',
    pruner=threshold_pruner, sampler=TPE_sampler
)

GPS_sampler =  optuna.samplers.GPSampler()
search_threshold_GPS = optimize(
    classifier, dataset.data, dataset.target, params,
    study_name='KNeighbors classifier, threshold pruner, GPS sampler',
    pruner=threshold_pruner, sampler=GPS_sampler
)

[I 2024-12-18 16:31:48,679] A new study created in memory with name: KNeighbors classifier
<ipython-input-10-dcaff464f85d>:6: ExperimentalWarning: OptunaSearchCV is experimental (supported from v0.17.0). The interface can change in the future.
  optuna_search = optuna.integration.OptunaSearchCV(
[I 2024-12-18 16:31:49,080] Trial 0 finished with value: 0.980399061032864 and parameters: {'n_neighbors': 66, 'weights': 'uniform', 'p': 13}. Best is trial 0 with value: 0.980399061032864.
[I 2024-12-18 16:31:49,347] Trial 1 finished with value: 0.9383802816901408 and parameters: {'n_neighbors': 1, 'weights': 'uniform', 'p': 17}. Best is trial 0 with value: 0.980399061032864.
[I 2024-12-18 16:31:49,625] Trial 2 finished with value: 0.9440140845070422 and parameters: {'n_neighbors': 4, 'weights': 'distance', 'p': 7}. Best is trial 0 with value: 0.980399061032864.
[I 2024-12-18 16:31:49,887] Trial 3 finished with value: 0.9411189358372457 and parameters: {'n_neighbors': 3, 'weights': 'uniform', 

CPU times: user 1min 59s, sys: 6.85 s, total: 2min 6s
Wall time: 2min 21s


### Предсказания

Теперь получим значения метрики recall для всего датаета для всех полученных оптимизаций

In [8]:
def get_results(
    search, X, y, title, metric=recall_score, metric_name='recall', results=pd.DataFrame()
):
    current = pd.DataFrame({
        metric_name: [metric(y, search.predict(X))]
    }, index=[title])
    results = pd.concat((results, current))
    return results

In [12]:
zipped = zip(
    (
        search, search_percentile_random,search_percentile_TPE,
        search_percentile_GPS, search_threshold_random, search_threshold_TPE,
        search_threshold_GPS
    ),
    (
        'No pruner, no sampler', 'Percentile pruner, random sampler',
        'Percentile pruner, TPE sampler', 'Percentile pruner, GPS sampler',
        'Threshold pruner, random sampler', 'Threshold pruner, TPE sampler',
        'Threshold pruner, GPS sampler'
    )
)

results = pd.DataFrame()

for local_search, title in zipped:
    results = get_results(
        local_search, dataset.data, dataset.target, title, results=results
    )

results

recall
No pruner, no sampler              0.974790
Percentile pruner, random sampler  0.974790
Percentile pruner, TPE sampler     0.980392
Percentile pruner, GPS sampler     0.977591
Threshold pruner, random sampler   1.000000
Threshold pruner, TPE sampler      0.977591
Threshold pruner, GPS sampler      0.977591

Итак, лучшие значения recall в 0.980392 кобинация без сэмплера и прунера

### Визуализация

Для примера визуалиризуем некоторые статистики у оптимизации без прунера и сэмплера

In [13]:
plot_slice(search.study)

Можно сделать следующие выводы:
- Чем выше количество соседей, тем выше значение метрики recall
- Гиперпараметр p тяготеет к большим значениям (от 10 до 15)
- На удивление большее значение метрики recall дают как раз невзвешенные соседи  

In [14]:
plot_optimization_history(search.study)

На данном рисунке можно видеть развитие процесса оптимизации в зависимости от номера итерации

In [15]:
plot_param_importances(search.study)

Самый важный гиперпараметр - количество учитываемых соседей, другие практически не оказывают никакого влияния

In [16]:
plot_contour(search.study)

Есть некая корреляция между параметром p и количество учитываемых соседей

In [17]:
plot_parallel_coordinate(search.study)

Для наибольших значений метрики recall харакетрны следующие комбинации гиперпараметров:
- Высокое значение ближайших соседей (от 50 до 100)
- Большое значение p (от 10 до 18)
- В большей степени невзвешенные соседи

### Выводы

В данной лабораторной работе я попробовал применить библиотеку Optuna для очередной задачи (с Optun-ой я уде был знаком до этого), а также узнал, как выводить логи процесса оптимизации в базу данных PostgreSQL